### PR TITLE
[SOL-72] Get current estimated fee in USD

### PR DESCRIPTION
**Pull request by gitteri**
This PR adds an endpoint and functionality to get the current fee in USD for a simple SPL token transfer.

### CODE REVIEW

**Comment by kilogold** [comment]
Heads up. #8 may introduce merge conflicts.

I'm available to help navigate them. Ping me

File: src/logic/fee.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
+import Decimal from 'decimal.js';
+import { RpcService } from '@/services/rpcService';
+import { supportedMints } from '@/app/config/mint';
+
+const convertLamportsToUSD = async (lamports: bigint): Promise<string> => {
```
Comment (kilogold) [inline src/logic/fee.ts:6]: Design question: Should the conversion be to the given (stablecoin) mint, instead of $USD?

File: tsconfig.json [hunk 2]
```diff
@@ -1,5 +1,6 @@
 {
   "compilerOptions": {
+    "target": "ES2020",
```
Comment (kilogold) [inline tsconfig.json]: I know nothing. Is this backwards compatible?
Does it unlock anything?

File: src/services/rpcService.ts [hunk 3]
```diff
@@ -1,134 +1,211 @@
...
+   */
+  async getSplBalance(publicKey: PublicKey, mint: PublicKey, programId = TOKEN_PROGRAM_ID): Promise<bigint> {
+    try {
+      const pk = new PublicKey(publicKey);
+      if (!validatePublicKey(pk)) {
+        return BigInt(0);
+      }
+
+      console.log("getting spl balance", mint);
+      const response = await this.connection.getTokenAccountsByOwner(
```
Comment (kilogold) [inline src/services/rpcService.ts]: We're following the Associated Token Account convention, so there's only ever one token account per mint per user.

I recommend using [this approach](https://github.com/solana-labs/solana-program-library/blob/d72289c79a04411c69a8bf1054f7156b6196f9b3/token/js/src/actions/getOrCreateAssociatedTokenAccount.ts#L51) to fetch the token account.

File: src/services/rpcService.ts [hunk 4]
```diff
@@ -1,134 +1,211 @@
...
+        console.log("No token account found for this wallet");
+        return BigInt(0);
       }
+
+      // The balance is stored in the last 8 bytes of the account data
+      // The bytes are stored in little-endian format
+      const tokenAccountInfo = response.value[0].account.data;
+      const balance = BigInt('0x' + Buffer.from(tokenAccountInfo.subarray(64, 72)).reverse().toString('hex'));
+
+      console.log("Debug: SPL balance for", pk.toBase58(), "is", balance.toString());
```
Comment (kilogold) [inline src/services/rpcService.ts]: You can get the token balance with an RPC call:
https://solana.com/docs/rpc/http/gettokenaccountbalance

File: src/services/rpcService.ts [hunk 5]
```diff
@@ -1,134 +1,211 @@
...
+        new PublicKey(mint.address),
+        toTokenAccount,
+        new PublicKey('11111111111111111111111111111111'), // Dummy owner public key
+        1, // Minimal amount for transfer
+        mint.decimals,
+        undefined,
+        TOKEN_PROGRAM_ID, // TODO: support token 2022
+      )
+    );
+    transaction.feePayer = new PublicKey('11111111111111111111111111111111'); // Dummy fee payer public key
```
Comment (kilogold) [inline src/services/rpcService.ts]: Consider making a constant "DummyPublicKey".

**Review by kilogold — CHANGES_REQUESTED** [review]

File: src/logic/fee.ts (hunk 1 above)
Comment (gitteri) [reply src/logic/fee.ts:6]: know a good public API to get that?

File: src/services/rpcService.ts (hunk 4 above)
Comment (gitteri) [reply src/services/rpcService.ts]: This doesn't support token 2022?

Comment (gitteri) [reply src/services/rpcService.ts]: ^^ actual question. I just don't see any docs around this

Comment (gitteri) [reply src/services/rpcService.ts]: Looks like it works. read through some old PRs in the rpc client

File: tsconfig.json (hunk 2 above)
Comment (gitteri) [reply tsconfig.json]: Looks like it does some things. I did this originally to add some bigint features that weren't in the default version. https://www.learningtypescript.com/articles/why-increase-your-tsconfig-target

File: src/logic/fee.ts (hunk 1 above)
Comment (kilogold) [reply src/logic/fee.ts:6]: I do not. I'll defer to your judgement

**Review by kilogold — APPROVED** [review]
Ship it

**Pull request merged**